homeassistant/components/pglab/__init__.py

```diff
@@ -0,0 +1,87 @@
+"""PG LAB Electronics integration."""
+from __future__ import annotations
+
+from pypglab.mqtt import Client, Sub_State, Subcribe_CallBack
+
+from homeassistant.components import mqtt
+from homeassistant.components.mqtt import ReceiveMessage
+from homeassistant.components.mqtt.subscription import (
+    async_prepare_subscribe_topics,
+    async_subscribe_topics,
+    async_unsubscribe_topics,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+
+from .const import (
+    DEVICE_ALREADY_DISCOVERED,
+    DISCONNECT_COMPONENT,
+    DISCOVERY_INSTANCE,
+    PLATFORMS,
+)
+from .discovery import CreateDiscovery
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up PG LAB Electronics integration from a config entry."""
+
+    # define the call back for pglab  module to publish a mqtt message
+    async def mqtt_publish(
+        topic: str, payload: str, qos: int | None = 0, retain: bool | None = False
+    ) -> None:
+        await mqtt.async_publish(hass, topic, payload, qos, retain)
+
+    # define the call back for pglab module to subscribe to a mqtt message
+    async def mqtt_subscribe(
+        sub_state: Sub_State, topic: str, callback_func: Subcribe_CallBack
+    ) -> Sub_State:
+        async def discovery_message_received(msg: ReceiveMessage) -> None:
+            callback_func(msg.topic, msg.payload)
+
+        topics = {
+            "pglab_subscribe_topic": {
+                "topic": topic,
+                "msg_callback": discovery_message_received,
+            }
+        }
+
+        sub_state = async_prepare_subscribe_topics(hass, sub_state, topics)
+        await async_subscribe_topics(hass, sub_state)
+        return sub_state
```

Comment:
These functions are internal MQTT methods, please only use imports from `homeassistant.components.mqtt`

Reply:
Tasmota also uses the subscription helper, I think that should be allowed. Maybe we should export them explicitly in `mqtt/__init__.py`

Reply:
https://github.com/home-assistant/core/pull/116150

Reply:
I suppose my code is fine considering emontnemery comment.
In general to make this new integration I used Tasmota as reference

Reply:
You need to change the code

Reply:
Sorry, can you be more specific ?  I'm not sure of how to change it

Reply:
This is fixed by importing ReceiveMessage directly from mqtt 👍